clash-lib/src/Clash/Core/Evaluator/Semantics.hs

```diff
@@ -0,0 +1,548 @@
+  * evaluatePrimWith (primitives only)
+  * evaluateAppWith  (both)
+
+The reason for this is that data constructors and primitives are only
+considered values when they are fully applied, so they include all arguments
+when evaluated. When evaluating a term, constructors and primitives will appear
+either on their own or nested in a sequence of App / TyApp nodes.
+
+The special functions, evaluateDataWith and evalautePrimWith handle the first
+case. As these are leaves in the term AST (they have no subterms), they are
+only values if they take no arguments, otherwise they are eta-expanded, and
+the newly created binders are used to provide variables for each argument.
+
+To identify sequences of applications which are applied to constructors or
+primitives, the evaluateAppWith function unwraps all applications. If it is
+a constructor or primitive, it checks that all arguments are applied and
+eta-expands if any are missing.
+
+There is one more consideration which applies to primitives only: the result
+of a primitive may potentially be a function (e.g. primtives that return an
+action in IO). This is handled in evaluateAppWith, by checking if more
+arguments are given than the primitive requires, and partitioning the
+arguments. The arguments for the primitive are applied and the primtive is
+evaluated, then the remaining arguments are applied.
+-}
+
+isFullyApplied :: Type -> [Either Term Type] -> Bool
+isFullyApplied ty args =
+  List.equalLength args (fst $ splitFunForallTy ty)
+
+etaExpand :: Term -> Eval Term
+etaExpand x =
+  case collectArgs x of
+    y@(Data dc, _) -> do
+      tcm <- State.gets envTcMap
+      expand tcm (dcType dc) y
+
+    y@(Prim p, _) -> do
+      tcm <- State.gets envTcMap
+      expand tcm (primType p) y
+
+    _ -> pure x
+ where
+  expand :: TyConMap -> Type -> (Term, [Either Term Type]) -> Eval Term
+  expand tcm ty (tm, args) = do
+    let missingArgTys = fst $ splitFunForallTy (applyTypeToArgs tm tcm ty args)
+    missingArgs <- traverse etaNameOf missingArgTys
+
+    pure $ mkAbstraction
+      (mkApps x (fmap (bimap Var VarTy) missingArgs))
+      missingArgs
+
+  etaNameOf :: Either TyVar Type -> Eval (Either Id TyVar)
+  etaNameOf = \case
+    Left tv  -> pure (Right tv)
+    Right ty -> Left <$> State.state (mkUniqueId "eta" ty)
+
+-- | Default implementation for looking up a variable in the environment.
+-- This checks both the local and global environments, and inlines global
+-- binders which are not marked as NOINLINE.
+--
+-- As variables can point to unevalated terms or WHNF values, this function
+-- also forces unevaluated terms to WHNF when performing lookups. This is
+-- currently performed in an environment without the given identifier, to
+-- prevent expressions like
+--
+--   let x = x in ...
+--
+-- looping if x is looked up, as it exists in the environment as
+--
+--   Var x |-> Var x
+--
+-- TODO: In the future this should not delete identifiers from the environment
```

Comment:
Ah, I guess this is were the small-step and big-step differ. In the small step, evaluating a variable (global or local) would put an Update frame on the stack, so once the variable was evaluated to a value (e.g. a Lambda), the heap/environment would be expanded with the variable-value combo again. And so, deleting functions from the heap would no prevent e.g. functions like fmap to be unrolled. Could something similar be done for the big-step version?

Reply:
The small step evaluator (as implemented in Clash) actually behaves in a similar way, unrolling things like `fmap` once. Consider the topentity

```haskell
topEntity :: [Int]
topEntity = fmap (*2) [1..10]
```

using the `partial-eval` debugging program in this branch (with the `experimental-evaluator` flag off it uses the small-step evaluator), you can see that this doesn't unroll all the way to

```haskell
topEntity :: [Int]
topEntity = [2,4,6,8,10,12,14,16,18,20]
```

but only unrolls one step giving something similar to [1]

```haskell
topEntity 2 : let wild = 1 in case wild of
                                10 -> []
                                _  -> topEntity_go (wild + 1)
```

where `topEntity_go` is a generated worker. For it to unroll all the way, it would need the `split` function from the paper it's based on to supercompile subterms, and ideally the `terminates` function from the same paper to prevent divergence.

As far as big-step / small-step differences go: the `quote` function is akin to their `split` function, so nothing to be done there. As already mentioned in #950, I want a termination checker for the evaluator too (but _not_ one using the tag bags / well-quasi-orderings from the small-step paper).

---

[1] I actually noticed while double checking this that the old evaluator gives a slightly incorrect result too, the binder `x` from `topEntity_go = \x -> ...` isn't fully substituted in the unrolled `topEntity`, appearing twice as a free variable, more like:

```haskell
topEntity = (I# (x *# 2)) : let wild = x in case wild of
                                10 -> []
                                _  -> topEntity_go (x + 1)
```

while the new evaluator substitutes in the value of `x` correctly.

Reply:
Right, it only unrolls once, because it only evaluates to WHNF; not NF. That has nothing to do with removing values form the heap while evaluating them to WHNF.